### PR TITLE
Fix available memory extraction on Linux

### LINKED ISSUE #26568: GC does not correctly evaluate the memory load on Linux
On Linux, when running in an unrestricted environment, the GC uses `sysconf(SYSCONF_PAGES) * sysconf(_SC_PAGE_SIZE)` to evaluate the total memory consumption of the system (https://github.com/dotnet/coreclr/blob/master/src/pal/src/misc/sysinfo.cpp#L368). 

`SYSCONF_PAGES` is mapped on `_SC_AVPHYS_PAGES`. Unfortunately, it counts the memory used by the page cache (which is automatically freed by the OS as needed), and therefore overestimates the system load. 

```
$ free -h
              total        used        free      shared  buff/cache   available
Mem:            62G         26G         30G        1.5M        6.1G         35G
Swap:            0B          0B          0B

$ getconf _AVPHYS_PAGES
7968847

$ getconf PAGESIZE
4096
```

We can see here that _AVPHYS_PAGES * PAGESIZE is 32 GB, even though only 26 GB of resident memory is actually used. We've seen instances where the GC incorrectly concludes that more of 90% of the memory is used, and start doing blocking collections even though it shouldn't be needed.

### PR DESCRIPTION
**Pull request by janvorli**
The GlobalMemoryStatusEx in PAL is returning number of free physical pages in
the ullAvailPhys member. But there are additional pages that are allocated
as buffers and caches that get released when there is a memory pressure and
thus they are effectively available too.

This change extracts the available memory on Linux from the /proc/meminfo
MemAvailable row, which is reported by the kernel as the most precise
amount of available memory.

Close #26568

### CODE REVIEW

**Comment by janvorli** [comment]
cc: @kevingosse, @Maoni0, @sergiy-k

**Comment by jkotas** [comment]
Do we need a matching change for standalone GC PAL?

File: src/pal/src/misc/sysinfo.cpp [hunk 1]
```diff
@@ -239,6 +240,58 @@ GetSystemInfo(
...
+            {
+                uint64_t multiplier = GetMemorySizeMultiplier(units);
+                *memAvailable = available * multiplier;
+                foundMemAvailable = true;
+                break;
+            }
+        }
+
+        free(line);
+        fclose(memInfoFile);
```
Comment (jkotas) [inline src/pal/src/misc/sysinfo.cpp]: Is `fclose` guaranteed to be robust against NULL argument?

Comment (janvorli) [reply src/pal/src/misc/sysinfo.cpp]: We don't call it on NULL argument

Comment (jkotas) [reply src/pal/src/misc/sysinfo.cpp]: Sorry, I have misread the indentation

**Comment by janvorli** [comment]
> Do we need a matching change for standalone GC PAL?

Yes, we will need it. In fact, I've found that the standalone GC uses a completely different way to get the physical memory used, which is in fact even more incorrect (it is based on memory consumed by the current process instead of the system one). I'll extend this PR with the standalone GC fix.

File: src/pal/src/misc/sysinfo.cpp [hunk 2]
```diff
@@ -239,6 +241,58 @@ GetSystemInfo(
...
+        char *line = nullptr;
+        size_t lineLen = 0;
+
+        while (getline(&line, &lineLen, memInfoFile) != -1)
+        {
+            char units = '\0';
+            uint64_t available;
+            int fieldsParsed = sscanf(line, "MemAvailable: %" SCNu64 " %cB", &available, &units);
+
+            if (fieldsParsed >= 1)
```
Comment (sergiy-k) [inline src/pal/src/misc/sysinfo.cpp]: Should this be >=2 since we need to parse values for 2 fields?

Comment (janvorli) [reply src/pal/src/misc/sysinfo.cpp]: No, since the units fields can be missing. So we will either get just the "available" or "available" and "units".

**Comment by janvorli** [comment]
I still need to test the standalone GC version, I just needed to push it so that I can fetch the change from my various distros / OS-es. So I am marking it as no merge until I verify that it works as expected on both Linux and OSX.

**Comment by Maoni0** [comment]
have you done the perf check with the GC perf infra? @andy-ms added the linux container support

**Comment by janvorli** [comment]
> have you done the perf check with the GC perf infra

Is there any doc on how to do that?

**Comment by andy-ms** [comment]
@janvorli I'll message you on Teams.

File: src/gc/unix/gcenv.unix.cpp [hunk 3]
```diff
@@ -722,6 +751,58 @@ static size_t GetLogicalProcessorCacheSizeFromOS()
...
+    return 1;
+}
+
+#ifndef __APPLE__
+// Try to read the MemAvailable entry from /proc/meminfo.
+// Return true if the /proc/meminfo existed, the entry was present and we were able to parse it.
+static bool ReadMemAvailable(uint64_t* memAvailable)
+{
+    bool foundMemAvailable = false;
+    FILE* memInfoFile = fopen("/proc/meminfo", "r");
```
Comment (lpereira) [inline src/gc/unix/gcenv.unix.cpp]: While this is fine -- you could also use `sysinfo(2)` here, which obtains the information from the same place that `procfs` uses to populate `/proc/meminfo`, but should work where `procfs` isn't available for whatever reason.

Comment (lpereira) [reply src/gc/unix/gcenv.unix.cpp]: Oh, nevermind, saw it's being used down the line. :)

File: src/pal/src/misc/sysinfo.cpp [hunk 4]
```diff
@@ -239,6 +241,58 @@ GetSystemInfo(
...
+        case 'M': return 1024 * 1024;
+        case 'k':
+        case 'K': return 1024;
+    }
+
+    // No units multiplier
+    return 1;
+}
+
+#ifndef __APPLE__
```
Comment (lpereira) [inline src/pal/src/misc/sysinfo.cpp]: Isn't it better to use platform-specific #ifdefs here? This should be built on FreeBSD as well, where it's going to fail during runtime (and when it's time to add support to other OSes, it's easier to have the preprocessor #error on an #else).

File: src/gc/unix/gcenv.unix.cpp (hunk 3 above)
Comment (janvorli) [reply src/gc/unix/gcenv.unix.cpp]: Actually, we use sysinfo only to extract the swap size. But looking at the sysinfo, it doesn't seem to include the equivalent of the MemAvailable.

File: src/pal/src/misc/sysinfo.cpp (hunk 4 above)
Comment (janvorli) [reply src/pal/src/misc/sysinfo.cpp]: It seems that procfs exists on other Unix systems too. And here we just try once and then fall back to using sysconf, so we don't fail.

**Pull request merged**